man/delta.1
.so .bitmacros
.TH DELTA 1 "May, 1998" "BitMover, Inc." BitSCCS
.SH NAME
delta \- make a delta to an SCCS file
.SH SYNOPSIS
.B /usr/bitsccs/delta
[\fB\-cilnpqs\fP]
[\fB-I\fIf\fR]
[\fB-R\fIr\fR]
[\fB-y\fIc\fR]
[\fIfile\fP \|.\|.\|.]
.SH DESCRIPTION
This command is part of \*[BS], the \*[BM] version of \*[SCCS].  For an 
introduction, please see
.BR bitsccs (1).
.LP
.B delta
checks in changes (aka deltas) in the working file to the
\*[SCCS] revision control file.
The working file[s] must have been previously edited by the
.BR get ,
.BR co ,
or some other \*[SCCS] enabled program.
.B delta
may prompt you on the standard output depending upon
the options specified and the flags that are set in the
.BR s. file
(see
.BR admin (1),
and the
.B -y
option below, for details).
.so filenames
.SH OPTIONS
.TP
.B -c
Do not regenerate the \*[SCCS] checksum after writing the file.
Use of this option is not advised, it is usually used  in conversion
scripts for a little extra performance.
.TP
.B -i
Initialize the file.  Used when checking in an initial file.  The traditional
\*[SCCS] command for this function is
.BR admin .
.TP
.TP
.BI -I initFile
Take the date, user, comments, pathname, hostname, etc., from the file
.IR initFile .
This option is typically used in conversion scripts.  The file format 
must match that produced by a
.B "prs -I<rev>" 
and looks like:
.DS
D 1.3 97/05/25 21:59:47 lm@bitmover.com 3 2  1/1/32
P /home/lm/smt/sccs/delta.c
C Support -y flag.
.DE
.TP
.B -l
Lock the file after the check in.  This is analogous to following the 
.B delta 
command with a 
.BR "get -e" .
.TP
.B -n
Retain the edited
.BR g -file,
which is normally removed at the completion of processing.
.TP
.B -p
Display differences (in
.BR diff (1)
format) on the standard output.
.TP
.B \-q
Run silently, like 
.B -s
below.
.TP
.BI \-R rev
Force the revision of the new delta to be
.IR rev .
Use this with caution, because an incorrect revision number can confuse
programs which parse the files.  This option is typically only used in
conversion scripts.
.TP
.B \-s
Run quietly.  All informational messages are suppressed; error messages
are unaffected.
.TP
.BI \-y [comment]
Supply a comment for the delta.  A
null
comment is acceptable.  If
.B \-y
is not specified,
.B delta
prompts with the file name and new revision number like so:
.DS
$ delta foo.c
End comments with "." by itself, blank line, or EOF.
foo.c 2.11>>  
.DE
.SH EXAMPLES
To check in everything in the local directory, try:
.DS
$ delta
.DE
If you wanted to use the same comment on every delta, you could also
try:
.DS
$ delta -y'Changes for release 1 of Widget'
.DE
A common usage is to print the differences, this is handy since you frequently
need to scan the differences to provide meaningful comment.  For each file,
the differences are printed, then you are prompted for the comment.
.DS
$ delta -p foo.c bar.c
===== foo.c 1.1 vs edited =====
1a
> /* Copyright 1998 Foo Corp, Ltd. */
foo.c 1.2>> Add copyright
foo.c 1.2>> .
===== bar.c 1.3 vs edited =====
1a
> /* Copyright 1998 Foo Corp, Ltd. */
bar.c 1.4>> Add copyright
bar.c 1.4>> .
.DE
.SH FILES
.TP 20
.BI p. file
lock file for a checked-out version
.TP
.BI s. file
\*[SCCS] history file
.TP
.BI x. file
temporary copy of the
.BR s. file
.TP
.BI z. file
lock file
.SH "SEE ALSO"
.BR admin (1),
.BR sccs-ci (1),
.BR get (1),
.BR prs (1),
.BR rmdel (1),
.BR diff (1),
.BR diffs (1).


RELEASE-NOTES
============================================================================
Release notes for BitKeeper version 6.0.2 (released Oct 29, 2013)

    This is a bugfix release (one feature snuck in).

Bugfixes

  bk diffs
    - Fixed a bug where 'bk diffs -NL' did not list extra files.
    - bk diffs -s was not working because of the switch to -u as
      the default.
    - Function header matching in 'bk diffs -p' now works with C++
      names.

  bk difftool
    - When looking at diffs over different types of files, a binary
      file in the list would cause diffs for the next text file to not
      be displayed.

  bk gone
    - People would try and run 'bk gone foo.c' which is incorrect, the
      correct usage is:
	bk gone `bk log -r+ -nd:ROOTKEY:`
      bk gone now works harder to validate that the arguments passed
      are actually keys.

  Other bugfixes
    - Fix a bug in the BK6->BK5 compat code where some files were not
      converted correctly to the BK5 format.
    - Stale locks on windows were not always being cleaned immediately.
    - Ignore extra files in .bk which were causing problems in Mac OS.

Features

   bk annotate
    - MD5 keys are now supported in the annotation options by running
      either 'bk annotate -a5' or 'bk annotate -A5'.

============================================================================
Release notes for BitKeeper version 6.0.1 (released Aug 9, 2013)

    This is a bugfix release (a few features snuck in).

Bugfixes

  bk clone
    - 'bk clone -@' wasn't populating comps when the remote repository
      was a subset of the local baseline.
    - 'bk clone -r<rev> [--upgrade|--downgrade] repo localcopy' could
      fail in some cases, and has been fixed.
    - 'bk clone -r<rev> bk://host/repo localcopy' will now fail if the
      BK client version is bk-6.0 and the bkd is bk-6.0.1 or later.
      The solution is to upgrade the client.

  bk commit
    - 'bk commit -f' was still prompting for comments.
    - 'bk commit' had lost some status reporting; that has been
      restored.

  bk collapse
    - 'bk collapse' would leave components marked as pending; this caused
      pull to error with a "KEY MISMATCH" message.
    - 'bk fix/collapse' in a gate was only partially disabled, it would
      collapse in the product but would refuse to collapse in
      components.

  bk diffs
    - 'bk -Ucx diffs -N' would only print diffs for one new file;
      fixed to print diffs for all new files

  bk difftool
    - The file listing box on difftool resizes itself to be as large
      as the longest pathname shown (it used to truncate).
    - 'bk difftool -S -Lpath/to/component' is now supported.
      (-S == scan only the current component)
    - The 'next' button in bk difftool didn't come up highlighted when
      there was more than one file.
    - Better error reporting with -L when a single GCA doesn't exist.

  bk pull
    - A pull from a remote repository served by a bk version 5.x could fail
      if a component clone was part of the operation.  It works now.

  bk push
    - A push from a local repository served by a bk version 5.x to a remote
      repository that uses new storage format of bk-6 could fail if a component
      clone was part of the operation.  It works now.

  bk revtool
    - A particular sequence of clicks and launching difftool could
      make revtool unresponsive.

    - When launching revtool with a line number (bk revtoo +lno),
      the comments for the revision that added that line were not being shown.

  bk sfiles
    - In rare cases, a pending component would not be seen as pending
      ('bk -pA' did not list them but it does now).

  Misc
    - In rare cases, permissions in the .bk/SCCS directory could be set
      wrong and cause bk to fail.  The known case has been fixed and
      the problem is now repaired automatically.

    - In another permissions failure, when a user didn't have write
      permissions on the repository, commit would fail with an
      unhelpful error message.  The problem is now reported more
      clearly.

  Graphical tools
    - When resizing graphical tools quickly, the resize labels could
      get out of sync; they update more frequently now.

Features
    - bk difftool now supports -w/-b to ignore white space changes
      like 'bk diffs'.  There is toolbar control of this feature as
      well.

    - BitKeeper versions prior to 6.0 may now clone repositories in the new
      format; they will automatically be downgraded to a compatible format.

============================================================================
Release notes for BitKeeper version 6.0 (released Jul 05, 2013)

    The primary focus of the 6.0 effort is performance improvements.

New repository format for performance
    BK6 has a new file format designed for performance.

    The new repository format cannot be read by bk-5.x releases
    of BitKeeper, but clone can convert a new repository to a bk-5.x
    readable repository.

Compatibility
    New file formats are a cause for concern but we've done the extra
    work so you may try it without forcing everyone to upgrade. BK6 can
    pull/push the new format from/to bk5's format.

    To enable the the new, faster format:
  
	bk clone --upgrade bk5-repo bk6-repo
  
    To downgrade to be compatible with bk5 (or bk4 if it is a non-nested
    repo):

	bk clone --downgrade bk6-repo bk5-repo

    To allow clones/pulls/pushes from bk4 & bk5, serve the repositories
    with a bk6 bkd:

    	cd bk6-repo
	bk pull bk://bk6-server/bk5-repo
	bk push bk://bk6-server/bk5-repo
 
Performance improvements:
    In the examples below a non-nested test repository contained 55,000
    files with 282,000 changesets on an NFS server.  Both the server
    and client caches were dropped (made cold) before each operation.

    Note that none of the performance improvements depend on the use of
    a file system notifier.  We considered and rejected such an approach

src/delta.c
/*
 * Copyright 1997-2016 BitMover, Inc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "system.h"
#include "sccs.h"

int
fix_gmode(sccs *s, int gflags)
{
	ser_t	d;

	/*
	 * Do not fix mode of symlink target, the file may not be 
	 * under BK control.
	 */
	if (S_ISLNK(s->mode)) return (0);

	if ((gflags&GET_EDIT) && WRITABLE(s))  return (0);
	if (!(gflags&GET_EDIT) && !WRITABLE(s))  return (0);

	d = sccs_top(s);
	if (MODE(s, d)) s->mode = MODE(s, d);
	unless (gflags&GET_EDIT) {
		s->mode &= ~0222;	/* turn off write mode */
	}

	if (chmod(s->gfile, s->mode)) return (1);
	return (0);
}

private int
delta_trigger(sccs *s)
{
	char	*p;
	int	ret;
	char	here[MAXPATH];

	/* set up enviroment before the cd2root */
	p = proj_relpath(s->proj, s->gfile);
	safe_putenv("BK_FILE=%s", p);
	free(p);
	putenv("BK_EVENT=delta");

	/* pushd repo root */
	strcpy(here, proj_cwd());
	if (chdir(proj_root(s->proj))) {
		perror("to pre-delta trigger");
		exit (1);
	}

	ret = trigger("delta", "pre");

	/* popd */
	if (chdir(here)) {
		perror("from pre-delta trigger");
		exit (1);
	}
	putenv("BK_FILE=");
	return (ret);
}

/*
 * Remove dangling deltas from the file.
 */
private int
strip_danglers(char *name, u32 flags)
{
	char	*p, **revs = 0;
	int	i;
	sccs	*s;
	ser_t	d;
	FILE	*f;

	s = sccs_init(name, INIT_WACKGRAPH);
	assert(s);
	for (d = TABLE(s); d >= TREE(s); d--) {
		if (DANGLING(s, d)) revs = addLine(revs, strdup(REV(s, d)));
	}
	sccs_free(s);
	p = aprintf("bk -?BK_NO_REPO_LOCK=YES stripdel -%sdC -", (flags&SILENT) ? "q" : "");
	f = popen(p, "w");
	EACH(revs) {
		fprintf(f, "%s|%s\n", name, revs[i]);
	}
	freeLines(revs, free);
	if ((i = SYSRET(pclose(f)))) {
		fprintf(stderr, "%s failed\n", p);
		free(p);
		return (i);
	}
	free(p);
	s = sccs_init(name, INIT_WACKGRAPH);
	assert(s);
	sccs_renumber(s, (flags&SILENT)|INIT_WACKGRAPH);
	/*
	 * If we are preserving the monotonic flag but our parent didn't
	 * have it, make a note of that.
	 */
	d = sccs_top(s);
	if ((XFLAGS(s, d) & X_MONOTONIC) &&
	    !(XFLAGS(s, PARENT(s, d)) & X_MONOTONIC)) {
		char	**comments = 0;
		char	*c;

		if (HAS_COMMENTS(s, d)) {
			c = strdup(COMMENTS(s, d));
			chomp(c);
			comments = addLine(comments, c);
		}
		comments = addLine(comments, strdup("Turn on MONOTONIC flag"));
		comments_set(s, d, comments);
		freeLines(comments, free);
	}
	sccs_adminFlag(s, NEWCKSUM|ADMIN_FORCE);
	sccs_free(s);
	return (0);
}

int
delta_main(int ac, char **av)
{
	sccs	*s;
	int	iflags = 0;
	int	dflags = 0;
	int	gflags = 0;
	int	sflags = SF_GFILE|SF_WRITE_OK|SF_NOHASREVS;
	int	checkout = 0, ignorePreference = 0;
	int	c, rc;
	char	*didciFile = 0;
	char	*initFile = 0;
	char	*diffsFile = 0;
	char	*prog, *name;
	char	*encp = 0;
	char	*mode = 0;
	FILE	*diffs = 0;
	FILE	*init = 0;
	FILE	*didci = 0;
	pfile	pf = {0};
	int	dash, dbsort = 1, errors = 0, fire, dangling;
	off_t	sz;
	project	*locked_proj = 0;
	int	prefer_cfile = 0;
	char	here[MAXPATH];
	longopt	lopts[] = {
		{ "csetmark", 301 },
		{ "prefer-cfile", 302 },
		{ "did-ci;", 303 },
		{ 0, 0 }
	};

	prog = strrchr(av[0], '/');
	if (prog) prog++;
	else prog = av[0];
	if (streq(prog, "new") ||
	    streq(prog, "enter") || streq(prog, "add")) {
		dflags |= DELTA_NEWFILE;
		sflags |= SF_NODIREXPAND;
		sflags &= ~SF_WRITE_OK;
	} else if (streq(prog, "dbnew")) {
		dflags |= DELTA_NEWFILE | DELTA_DB;
		sflags |= SF_NODIREXPAND;
		sflags &= ~SF_WRITE_OK;
	}

	while ((c =
	    getopt(ac, av, "abcCdD:E|fGI;ilm|M;npPqRsSTuy|Y|Z|", lopts)) != -1) {
		switch (c) {
		    /* SCCS flags */
		    case 'n': dflags |= DELTA_SAVEGFILE; break;	/* undoc? 2.0 */
		    case 'p': dflags |= PRINT; break; 		/* doc 2.0 */
		    case 'm':					/* ci compat */
		    case 'y': 					/* doc 2.0 */
			if (comments_save(optarg)) return (1);
			dflags |= DELTA_DONTASK;
			break;
		    case 's': /* fall through */		/* undoc 2.0 */

		    /* RCS flags */
		    case 'q': 					/* doc 2.0 */
			trigger_setQuiet(1);
			dflags |= SILENT; gflags |= SILENT;
			break;
		    case 'f': dflags |= DELTA_FORCE; break;	/* doc 2.0 ci */
		    case 'i': dflags |= DELTA_NEWFILE; 		/* doc 2.0 */
			      sflags |= SF_NODIREXPAND;
			      sflags &= ~SF_WRITE_OK;
			      break;
		    case 'l': checkout |= CO_EDIT;		/* doc 2.0 */ 
			      break;
		    case 'u': checkout |= CO_GET;		/* doc 2.0 */
			      break;

		    /* LM flags */
		    case 'a':					/* doc 2.0 */
		    	dflags |= DELTA_AUTO;
			dflags &= ~DELTA_FORCE;
			sflags &= ~SF_WRITE_OK;
			break;
		    case 'b':	/* -b == -Ebinary */		/* doc 2.0 */
			if (streq(prog, "new")) {
		    		encp = "binary";
			} else {
				usage();
			}
			break;
		    case 'C': iflags |= INIT_NOCKSUM; break; 	/* undoc */
		    case 'c': dflags |= DELTA_CFILE; break;	/* doc */
		    case 'd': /* internal interface */ 		/* undoc 2.0 */
			      dflags |= DELTA_NOPENDING; break;
		    case 'D': diffsFile = optarg;		 /* doc 2.0 */
			      sflags = ~(SF_GFILE | SF_WRITE_OK);
			      break;
		    case 'I':
			initFile = optarg;
			dflags |= DELTA_FORCE;
			break;
		    case 'M': mode = optarg; break;		/* doc 2.0 */
		    case 'P': ignorePreference = 1;  break;	/* undoc 2.0 */
		    case 'R': dflags |= DELTA_PATCH; break;	/* undoc? 2.0 */
		    case 'G': 	/* remove -G in 2009 .  Replaced by -T */
		    case 'T': iflags |= INIT_FIXDTIME; break;	/* undoc? 2.0 */
		    case 'Y':
			if (optarg && comments_savefile(optarg)) {
				fprintf(stderr,
					"delta: can't read comments from %s\n",
					optarg);
				return (1);
			}
			dflags |= DELTA_DONTASK;
			break; 	/* doc 2.0 */
		    case 'Z': 					/* doc 2.0 */
			bk_setConfig("compression", optarg ? optarg : "gzip");
			break;
		    case 'E': encp = optarg; break; 		/* doc 2.0 */
		    case 'S': dbsort = 0; break;		/* undoc */
		    case 301:	// --csetmark
			dflags |= DELTA_CSETMARK; break;	/* undoc */
		    case 302:	// --prefer-cfile
			prefer_cfile = 1; break;		/* undoc */
		    case 303:	// --did-ci=
			didciFile = optarg;
			break;

		    default: bk_badArg(c, av);
		}
	}


	if (encp) {
		unless (dflags & DELTA_NEWFILE) {
			fprintf(stderr,
			    "Encoding is allowed only when creating files\n");
			usage();
		}

		/* check that they gave us something we can parse */
		if (sccs_encoding(0, 0, encp) == -1) usage();
	}

	if (chk_host() || chk_user()) return (1);

	dash = av[optind] && streq(av[optind], "-");
	name = sfileFirst(av[0], &av[optind], sflags);
	/* They can only have an initFile for one file...
	 * So we go get the next file and error if there
	 * is one.
	 */
	if ((initFile || diffsFile) && name && sfileNext()) {
		fprintf(stderr,
"%s: only one file may be specified with init or diffs file.\n", av[0]);
		usage();
	}

	/* force them to do something sane */
	if (!comments_got() &&
	    dash && name && !(dflags & (DELTA_NEWFILE|DELTA_CFILE)) && sfileNext()) {
		fprintf(stderr,
"%s: only one file may be specified without a checkin comment\n", av[0]);
		usage();
	}
	if (!comments_got() && prefer_cfile) {
		fprintf(stderr,
		    "%s: must have -y or -Y when using --prefer-cfile\n", prog);
		usage();
	}
	if (initFile && (dflags & DELTA_DONTASK)) {
		fprintf(stderr,
		    "%s: only init file or comment, not both.\n", av[0]);
		usage();
	}
	if ((checkout & CO_GET) && (checkout & CO_EDIT)) {
		fprintf(stderr, "%s: -l and -u are mutually exclusive.\n",
			av[0]);
		usage();
	}
	if (diffsFile && !(diffs = fopen(diffsFile, "r"))) {
		fprintf(stderr, "%s: diffs file '%s': %s.\n",
			av[0], diffsFile, strerror(errno));
	       return (1);
	}
	if (initFile && !(init = fopen(initFile, "r"))) {
		fprintf(stderr,"%s: init file '%s': %s.\n",
			av[0], initFile, strerror(errno));
		return (1);
	}
	if (didciFile && !(didci = fopen(didciFile, "w"))) {
		fprintf(stderr, "%s: didci file '%s': %s.\n",
		    av[0], didciFile, strerror(errno));
		return (1);
	}
	if ((fire = (getenv("_IN_DELTA") == 0))) putenv("_IN_DELTA=YES");
	strcpy(here, proj_cwd());
	
	while (name) {
		ser_t	d = 0;
		char	*nrev;
		int	df = dflags;
		int	gf = gflags;
		int	reget = 0;
		int	co;

		unless (s = sccs_init(name, iflags)) {
			name = sfileNext();
			errors |= 1;
			continue;
		}
		if (CSET(s)) {
			unless (dash) {
				fprintf(stderr, "%s: skipping ChangeSet\n",
				    prog);
			}
			goto next;
		}
		if (mode) d = sccs_parseArg(s, d, 'O', mode, 0);
		if (df & DELTA_AUTO) {
			if (HAS_SFILE(s)) {
				df &= ~DELTA_NEWFILE;
			} else {
				df |= DELTA_NEWFILE;
			}
		}
		/* DB files must be sorted unless specified otherwise. */
		if (DB(s)) df |= DELTA_DB;
		if ((df & DELTA_DB) && dbsort) {
			if (db_sort(s->gfile, s->gfile)) {
				fprintf(stderr,
				    "Error loading DB file %s, skipped\n",
				    s->gfile);
				errors |= 4;
				goto next;
			}
		}
		/*
		 * Checkout option does not applies to ChangeSet file
		 * see rev 1.118
		 * XXX - delta -l|u will be ignored for ChangeSet.
		 */
		co = checkout;		// This is from -l/-u, it goes first
		unless (co || ignorePreference) co = CO(s);
		unless (CSET(s)) {
			if (co & (CO_EDIT|CO_LAST)) {
				gf |= GET_SKIPGET|GET_EDIT;
				df |= DELTA_SAVEGFILE;
				reget = 1;	/* to write pfile */
			} else if (co & CO_GET) {
				if (HAS_KEYWORDS(s))  {
					gf |= GET_EXPAND;
					reget = 1;
				} else {
					gf |= GET_SKIPGET;
					df |= DELTA_SAVEGFILE;
				}
			}
		}

		nrev = NULL;
		if (HAS_PFILE(s)) {
			if (sccs_read_pfile(s, &pf)) {
				errors |= 2;
				goto next;
			}
			nrev = pf.newrev;
		}
		if (df & DELTA_DONTASK) {
			unless (d = comments_get(s->gfile, nrev, s, d)) {
				errors |= 1;
				break;
			}
			if (prefer_cfile) comments_readcfile(s, 0, d);
		}
		if (fire && proj_root(s->proj) &&
		    (locked_proj != s->proj)) {
			if (locked_proj) {
				chdir(proj_root(locked_proj));
				cmdlog_unlock(CMD_WRLOCK);
				proj_free(locked_proj);
			}
			chdir(proj_root(s->proj));
			cmdlog_lock(CMD_WRLOCK);
			locked_proj = proj_init(proj_root(s->proj));;
			chdir(here);
		}
		if (fire && proj_hasDeltaTriggers(s->proj)) {
			win32_close(s);
			switch (delta_trigger(s)) {
			    case 0:
				break;
			    case 2: /* trigger ran delta, we won't */
				goto next;
			    default:
				errors |= 32;
				goto next;
			}
		}

		if (df & DELTA_NEWFILE) {
			sz = diffs ? size(diffsFile) : size(s->gfile);
			s->encoding_in = s->encoding_out =
			    sccs_encoding(s, sz, encp);
		}

		dangling = MONOTONIC(s) && DANGLING(s, sccs_top(s));
		if (dangling) df |= DELTA_MONOTONIC;
		rc = sccs_delta(s, df, d, init, diffs, 0);
		if (diffs) fclose(diffs);
		if (rc == -4) {	/* interrupt in comment prompt */
			errors |= 4;
			break;
		}
		if (rc == -2) goto next; /* no diff in file */
		if (rc == -1) {
			sccs_whynot("delta", s);
			errors |= 4;
			goto next;
		}

		if (didci) {
			fprintf(didci, "%s|%s\n",
			    s->gfile, REV(s, sccs_top(s)));
		}

		if (dangling) {
			sccs_free(s);
			strip_danglers(name, dflags);
			s = sccs_init(name, iflags);
			assert(s);
			d = TABLE(s);
			assert(d);
			assert(!TAG(s, d));
		} else {
			s = sccs_restart(s);
		}
		unless (s) {
			fprintf(stderr,
			    "%s: can't restart %s\n", av[0], name);
			errors |= 8;
			goto next;
		}
		if (df & DELTA_SAVEGFILE) {
			/*
			 * If delta -Ddiffs with no gfile, then no GFILE,
			 * and we'd like to get one.
			 *
			 * fix_gmode() will fail if we don't own
			 * the gfile, in that case set the stage
			 * to re-get it
			 */
			if (!HAS_GFILE(s) || fix_gmode(s, gf)) {
				reget = 1;
				gf &= ~GET_SKIPGET;
			}
		}
		if (df & DELTA_NEWFILE) {
			/*
			 * The 'keyword' preference for a new file might
			 * have set keywords in sccs_delta() so we need to
			 * check again.
			 */
			if (HAS_KEYWORDS(s) && !(gf & GET_EDIT)) {
				gf |= GET_EXPAND;
				gf &= ~GET_SKIPGET;
				reget = 0;
				unless (co = checkout) co = CO(s);
				reget = co & CO_GET;
			}
		}

		if (reget) {
			// XXX - what if we are dangling?
			// The pf.oldrev is definitely wrong.
			nrev = (rc == -3) ? pf.oldrev : REV(s, TABLE(s));
			if (sccs_get(s, nrev, 0, 0, 0, gf, s->gfile, 0)) {
				unless (BEEN_WARNED(s)) {
					fprintf(stderr,
					    "get of %s failed, skipping it.\n",
					    name);
				}
			}
		}
next:		if (init) fclose(init);
		/*
		 * No, sccs_diffs() does this.
		 * if (diffs) fclose(diffs);
		 */
		sccs_free(s);
		free_pfile(&pf);
		unless (df & DELTA_DONTASK) comments_done();
		name = sfileNext();
	}
	if (didci) fclose(didci);
	if (sfileDone()) errors |= 64;
	if (locked_proj) {
		chdir(proj_root(locked_proj));
		cmdlog_unlock(CMD_WRLOCK);
		proj_free(locked_proj);
		chdir(here);
	}
	comments_done();
	return (errors);
}
